internal/telemetry/telemetry.go

```diff
@@ -0,0 +1,167 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"contrib.go.opencensus.io/exporter/ocagent"
+	datadog "github.com/DataDog/opencensus-go-exporter-datadog"
+	"github.com/hashicorp/go-hclog"
+	ocview "go.opencensus.io/stats/view"
+	octrace "go.opencensus.io/trace"
+	"go.opencensus.io/zpages"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Option func(*telemetry)
+
+func Run(opts ...Option) error {
+	var t telemetry
+	for _, opt := range opts {
+		opt(&t)
+	}
+
+	if t.Logger == nil {
+		t.Logger = hclog.L().Named("telemetry")
+	}
+	log := t.Logger
+
+	// If any of the below agents or servers need to close on exit, they can
+	// register their close function here, and they will be called when the parent
+	// context closes this runner.
+	var closeFuncs []func()
+
+	if t.EnableOpenCensusExporter {
+		log.Debug("Starting the opencensus agent exporter")
+		exporter, err := ocagent.NewExporter(t.OpenCensusExporterOptions...)
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "failed to initalize opencensus agent exporter: %s", err)
+		}
+		octrace.RegisterExporter(exporter)
+		ocview.RegisterExporter(exporter)
+		closeFuncs = append(closeFuncs, func() {
+			log.Debug("Shutting down opencensus agent exporter")
+			exporter.Flush()
+			if err := exporter.Stop(); err != nil {
+				log.Error("Failed to stop the opencensus agent exporter", "err", err)
+			} else {
+				log.Debug("OpenCensus agent exporter flushed and stopped")
+			}
+		})
+	}
+
+	if t.EnableDatadogExporter {
+		log.Debug("Starting the opencensus datadog exporter")
+		exporter, err := datadog.NewExporter(t.DatadogExporterOptions)
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "failed to initalize datadog exporter: %s", err)
+		}
+		octrace.RegisterExporter(exporter)
+		ocview.RegisterExporter(exporter)
+		closeFuncs = append(closeFuncs, func() {
+			log.Debug("Shutting down datadog exporter")
+			exporter.Stop()
+			log.Debug("Datadog exporter flushed and stopped")
+		})
+	}
+
+	// Run zPages
+	if t.EnableZpages {
+		zPagesMux := http.NewServeMux()
+		zpages.Handle(zPagesMux, "/debug")
+
+		var zpagesAddr string
+		if t.ZpagesAddr == "" {
+			zpagesAddr = "127.0.0.1:9999"
+		} else {
+			zpagesAddr = t.ZpagesAddr
+		}
+
+		srv := http.Server{
+			Addr:    zpagesAddr,
+			Handler: zPagesMux,
+		}
+
+		go func() {
+			log.Debug("Starting zPages server", "addr", zpagesAddr)
+			if err := srv.ListenAndServe(); err != nil {
+				log.Debug("zPages server exited", "err", err)
+			}
+		}()
+		closeFuncs = append(closeFuncs, func() {
+			err := srv.Close()
+			if err != nil {
+				log.Error(fmt.Sprintf("Failed to shut down zPages server: %s", err))
+			}
+		})
+	}
+
+	// Less frequent sampling can be achieved by exporting to an opencensus collector with sampling configured.
+	octrace.ApplyConfig(octrace.Config{DefaultSampler: octrace.AlwaysSample()})
+
+	// Wait on context to close
+	<-t.Context.Done()
```

Comment:
One reason to split them out as `Start()` and `Wait()` rather than just `Run()` is that you can report a startup error immediately and handle it during server startup, rather than having Run() return a startup error that is probably in it's own goroutine.